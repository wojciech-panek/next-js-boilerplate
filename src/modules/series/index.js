export {
  SeriesTypes,
  SeriesActions,
} from './series.redux';

export * from './series.selectors';
