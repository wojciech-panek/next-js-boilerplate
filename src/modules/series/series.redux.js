import { createActions, createReducer } from 'reduxsauce';
import { Record, List, Map, fromJS } from 'immutable';


export const { Types: SeriesTypes, Creators: SeriesActions } = createActions({
  fetch: [],
  fetchSuccess: ['data'],
  fetchError: ['payload'],
  fetchSingle: ['id'],
  fetchSingleSuccess: ['data'],
  fetchSingleError: ['payload'],
}, { prefix: 'SERIES_' });

const SeriesRecord = new Record({
  items: List(),
  data: Map(),
});

export const INITIAL_STATE = new SeriesRecord({});

const getSuccessHandler = (state = INITIAL_STATE, action) => state.set('items', fromJS(action.data));

const getSingleSuccessHandler = (state = INITIAL_STATE, action) => state.set('data', fromJS(action.data));

export const reducer = createReducer(INITIAL_STATE, {
  [SeriesTypes.FETCH_SUCCESS]: getSuccessHandler,
  [SeriesTypes.FETCH_SINGLE_SUCCESS]: getSingleSuccessHandler,
});
