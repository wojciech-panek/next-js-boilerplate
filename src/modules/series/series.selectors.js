import { createSelector } from 'reselect';


const selectSeriesDomain = state => state.get('series');

export const selectSeriesItems = createSelector(
  selectSeriesDomain, state => state.get('items')
);

export const selectSeriesData = createSelector(
  selectSeriesDomain, state => state.get('data')
);
