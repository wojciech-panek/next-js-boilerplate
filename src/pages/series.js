import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'next/router';
import { compose } from 'ramda';

import withIntl from '../shared/lib/withIntl';


export class SeriesPage extends PureComponent {
  static propTypes = {
    router: PropTypes.object.isRequired,
  };

  render() {
    return (
      <div>Series: {this.props.router.query.id}</div>
    );
  }
}

export default compose(
  withIntl,
  withRouter,
)(SeriesPage);
