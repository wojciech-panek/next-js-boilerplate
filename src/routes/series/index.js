export { default as Series } from './series.container';
