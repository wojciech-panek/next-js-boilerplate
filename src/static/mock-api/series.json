[
  {
    "id": 1,
    "title": "Dick in Ibiza",
    "season": "1"
  },
  {
    "id": 2,
    "title": "Dick in Ibiza",
    "season": "2"
  },
  {
    "id": 3,
    "title": "Liars",
    "season": "1"
  },
  {
    "id": 4,
    "title": "F*ck My Life",
    "season": "1"
  }
]
