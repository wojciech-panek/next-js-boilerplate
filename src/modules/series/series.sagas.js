import { put, takeLatest } from 'redux-saga/effects';
import { delay } from 'redux-saga';

import reportError from '../../shared/utils/reportError';
import api from '../../shared/services/api';
import { SeriesTypes, SeriesActions } from './series.redux';


export function* fetchSeries() {
  try {
    yield delay(1000);
    const { data } = yield api.get('/static/mock-api/series.json');

    return yield put(SeriesActions.fetchSuccess(data));
  } catch (e) {
    if (e.response) {
      return yield put(SeriesActions.fetchError(e.response.data));
    }

    return yield reportError(e);
  }
}

export function* fetchSingleSeries({ id }) {
  try {
    yield delay(1000);
    const { data: seriesData } = yield api.get('/static/mock-api/series.json');
    const { data: episodesData } = yield api.get('/static/mock-api/episodes.json');

    const data = seriesData.find((series) => series.id === id);
    data.episodes = episodesData.filter((episide) => episide.series === id);

    return yield put(SeriesActions.fetchSingleSuccess(data));
  } catch (e) {
    if (e.response) {
      return yield put(SeriesActions.fetchSingleError(e.response.data));
    }

    return yield reportError(e);
  }
}

export function* watchSeries() {
  try {
    yield takeLatest(SeriesTypes.FETCH, fetchSeries);
    yield takeLatest(SeriesTypes.FETCH_SINGLE, fetchSingleSeries);
  } catch (e) {
    yield reportError(e);
  }
}
