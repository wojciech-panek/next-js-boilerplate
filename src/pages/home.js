import React, { PureComponent } from 'react';
import { compose } from 'ramda';

import withIntl from '../shared/lib/withIntl';
import { Home } from '../routes/home';
import { SeriesActions } from '../modules/series/series.redux';


export class HomePage extends PureComponent {
  static fetchData(dispatch) {
    dispatch(SeriesActions.fetch());
  }

  static async getInitialProps({ isServer, store }) {
    if (isServer) {
      await store.execSagaTasks(isServer, (dispatch) => {
        HomePage.fetchData(dispatch);
      });
    } else {
      HomePage.fetchData(store.dispatch);
    }

    return {};
  }

  render() {
    return <Home />;
  }
}

export default compose(
  withIntl,
)(HomePage);
