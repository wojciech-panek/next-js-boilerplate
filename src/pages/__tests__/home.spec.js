import React from 'react';
import { shallow } from 'enzyme';
import { spy } from 'sinon';
import { expect } from 'chai';

import { HomePage } from '../home';
import { SeriesActions } from '../../modules/series/series.redux';


describe('Home: Page', () => {
  const defaultProps = {
  };

  const component = (props) => (
    <HomePage {...defaultProps} {...props} />
  );

  it('should render correctly', () => {
    const wrapper = shallow(component());
    global.expect(wrapper).toMatchSnapshot();
  });

  it('should dispatch SeriesActions.fetch in getInitialProps on server', async () => {
    const dispatch = spy();
    const store = global.utils.prepareInitialPropsStore(dispatch);

    await HomePage.getInitialProps({ isServer: true, store });

    expect(dispatch).to.have.been.calledOnce;
    expect(dispatch).to.have.been.calledWith(SeriesActions.fetch());
  });

  it('should dispatch SeriesActions.fetch in getInitialProps on client', async () => {
    const store = {
      dispatch: spy(),
    };

    await HomePage.getInitialProps({ isServer: false, store });

    expect(store.dispatch).to.have.been.calledOnce;
    expect(store.dispatch).to.have.been.calledWith(SeriesActions.fetch());
  });
});
