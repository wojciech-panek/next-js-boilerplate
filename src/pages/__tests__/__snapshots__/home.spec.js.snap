// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home: Page should render correctly 1`] = `<InjectIntl(withRouter(Connect(Home))) />`;
